packages/react/src/utils/floating-ui/useClick.ts

```diff
@@ -0,0 +1,126 @@
+   * determines whether to keep the floating element open when clicking the
+   * reference element for the first time.
+   * @default true
+   */
+  stickIfOpen?: boolean;
+}
+
+/**
+ * Opens or closes the floating element when clicking the reference element.
+ * @see https://floating-ui.com/docs/useClick
+ */
+export function useClick(context: FloatingRootContext, props: UseClickProps = {}): ElementProps {
+  const { open, onOpenChange, dataRef } = context;
+  const {
+    enabled = true,
+    event: eventOption = 'click',
+    toggle = true,
+    ignoreMouse = false,
+    stickIfOpen = true,
+  } = props;
+
+  const pointerTypeRef = React.useRef<'mouse' | 'pen' | 'touch'>(undefined);
+  const frameRef = React.useRef(-1);
+
+  React.useEffect(() => {
+    return () => {
+      if (frameRef.current !== -1) {
+        cancelAnimationFrame(frameRef.current);
+        frameRef.current = -1;
+      }
+    };
+  }, []);
+
+  const reference: ElementProps['reference'] = React.useMemo(
+    () => ({
+      onPointerDown(event) {
+        pointerTypeRef.current = event.pointerType;
+      },
+      onMouseDown(event) {
+        const pointerType = pointerTypeRef.current;
+        const nativeEvent = event.nativeEvent;
+
+        // Ignore all buttons except for the "main" button.
+        // https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/button
+        if (
+          event.button !== 0 ||
+          eventOption === 'click' ||
+          (isMouseLikePointerType(pointerType, true) && ignoreMouse)
+        ) {
+          return;
+        }
+
+        const nextOpen = !(
+          open &&
+          toggle &&
+          (dataRef.current.openEvent && stickIfOpen
+            ? dataRef.current.openEvent.type === 'mousedown'
+            : true)
+        );
+        // Wait until focus is set on the element. This is an alternative to
+        // `event.preventDefault()` to avoid :focus-visible from appearing when using a pointer.
+        frameRef.current = requestAnimationFrame(() => {
+          onOpenChange(nextOpen, nativeEvent, 'click');
+        });
+      },
+      onClick(event) {
+        const pointerType = pointerTypeRef.current;
+
+        if (eventOption === 'mousedown' && pointerType) {
+          pointerTypeRef.current = undefined;
+          return;
+        }
+
+        if (isMouseLikePointerType(pointerType, true) && ignoreMouse) {
+          return;
+        }
+
+        const nextOpen = !(
+          open &&
+          toggle &&
+          (dataRef.current.openEvent && stickIfOpen
+            ? dataRef.current.openEvent.type === 'click'
+            : true)
+        );
+        onOpenChange(nextOpen, event.nativeEvent, 'click');
+      },
+      onKeyDown() {
+        pointerTypeRef.current = undefined;
+      },
+    }),
+    [dataRef, eventOption, ignoreMouse, onOpenChange, open, stickIfOpen, toggle],
+  );
+
+  return React.useMemo(() => (enabled ? { reference } : {}), [enabled, reference]);
```

Comment:
Return a stable (const) object instead of `{}` to avoid recreating the return value when `enabled=false` and `reference` changing.